### PR TITLE
[Proposal] A pre-release review process for new major releases of any client library

### PR DESCRIPTION
**Pull request by bogdandrutu**
The main idea is that before every major release or the release of a new telemetry type (metrics)
we would like to do a technical and specification conformance review that will ensure consistency
and same "look and feel" for all officially released libraries.

### CODE REVIEW

**Comment by iNikem** [comment]
Can I ask why does TC think it needs such strict gating process?

**Review by alolita — COMMENTED** [review]
Great start on the TC helping regularize the release process. 

Some suggestions - 
* Add an overall purpose statement for this pre-release review process - what does this process aim to accomplish, outline benefits to the project.
* Add a checklist or list of questions maintainers can use for a spec compliance review. 
* Add a table to outline process milestones and timelines.

**Comment by tigrannajaryan** [comment]
> Can I ask why does TC think it needs such strict gating process?

A few reasons from the top of my head:
- To help language SIGs to find out whether they are ready for GA.
- To independently certify that language implementation is spec-compliant.
- To ensure uniform experience across languages (spec is useful, but does not fully capture usability aspects and how the API compares for different languages).
- To uncover gaps in the implementation (additional untrained pair of eyes often helps).

There is probably more.

**Review by jmacd — APPROVED** [review]
I think this will be great fort consistency! I'd like to add some kind of commitment to do these reviews swiftly, but could leave that out.

File: releases/new-telemetry-release-process.md [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+
+## Technical and Specification Compliance Review
+
+The OpenTelemetry [Technical Committee](https://github.com/open-telemetry/community/blob/main/tech-committee-charter.md)
+commits to ensure that every decision will be made in less than 14 days after
+the process is started.
+
+To begin the technical and specification compliance review language maintainers
+MUST submit a PR and complete this [template](new-telemetry-release-process-template.md).
+The name of the file MUST be `technical-specification-review-${language}.md`.
```
Comment (reyang) [inline releases/new-telemetry-release-process.md:36]: Would this cause name conflict? (e.g. `technical-specification-review-python.md` for tracing, and the same name for metrics later?)

Comment (lzchen) [reply releases/new-telemetry-release-process.md:36]: As well, what if there are multiple reviews done?

File: releases/new-telemetry-release-process-template.md [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+[compliance matrix](https://github.com/open-telemetry/opentelemetry-specification/blob/main/spec-compliance-matrix.md)
+for that language/telemetry.
+* Each language implementation MUST follow
+[the versioning and stability requirements](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/versioning-and-stability.md)
+and prepare the public documentation about the versioning and stability guarantees.
+* Work with a Technical Committee Sponsor(s) during review. The Technical
+Committee members are not language experts, and it is expected to work with the
+language maintainers and/or invited language experts to perform this review
+process. Language maintainers SHOULD respond to any question during the review
+time in a reasonable amount of time to not delay the review process.
```
Comment (reyang) [inline releases/new-telemetry-release-process-template.md:35]: Later in the doc we mentioned "every decision will be made in less than 14 days after the process is started", so there is no way a language maintainer could delay the process.

Perhaps we can rephrase this to "All the questions and feedbacks will be communicated as GitHub issues filed against the language client repo. If the language maintainers don't respond before the deadline (which is two weeks after the initial filing of the PR), the PR will be closed as REJECTED.".

File: releases/new-telemetry-release-process-template.md [hunk 3]
```diff
@@ -0,0 +1,48 @@
+# Technical Review and Specification Conformance for ${Language}/${TELEMETRY}/${MAJOR_VERSION} library
+
+Technical Committee Sponsor(s): []
+
+Status: [PROPOSAL|REJECTED|APPROVED]
```
Comment (reyang) [inline releases/new-telemetry-release-process-template.md:5]: Do we intend to merge these files anyways? If we only merge approved ones, there is no need to have `Status`.

File: releases/new-telemetry-release-process-template.md (hunk 2 above)
Comment (lzchen) [reply releases/new-telemetry-release-process-template.md:35]: Is 14days the deadline for the ENTIRE process (beginning of process, TC review, language contributors addressing/fixing change, new release), or just the deadline for the contributors' response?

File: releases/new-telemetry-release-process-template.md [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+
+Expected Release Date: []
+
+Spec Compliance Matrix is up to date: [YES|NO]
```
Comment (reyang) [inline releases/new-telemetry-release-process-template.md:9]: This will cause issue since other folks could add new requirements to the spec compliance matrix (unless we lock the spec during language client release). Consider putting a permanent link here pointing to a specific version of the spec compliance matrix.

File: releases/new-telemetry-release-process-template.md [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+
+Versioning And Stability Document: [LINK to version document]
```
Comment (reyang) [inline releases/new-telemetry-release-process-template.md:11]: We only need this once? (when metrics come, we don't need to redo this work)

File: releases/new-telemetry-release-process-template.md [hunk 6]
```diff
@@ -0,0 +1,48 @@
+# Technical Review and Specification Conformance for ${Language}/${TELEMETRY}/${MAJOR_VERSION} library
```
Comment (reyang) [inline releases/new-telemetry-release-process-template.md:1]: Need to clarify what `Telemetry` is, and (assuming it is "telemetry signal") what should people do when they have more than 1 signals in a single release.

File: releases/new-telemetry-release-process-template.md [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+It is highly recommended to have at least 2 members of the Technical Committee as sponsors. Any Technical Committee can be a sponsor, and any Technical Committee member can provide feedback during the review process.
+
+Technical Committee Sponsor(s) MUST:
+
+* Do the due diligence and review the public documentation and examples, and
+ensure specification conformance.
+* Ensure conformance with the versioning and stability document.
+* Ensure consistent names across implementations (e.g. TraceId vs Id)
+* Avoid confusions across implementation (e.g same function returns has different behaviors)
+* Ensure no experimental features (signals) are part of the main API package.
```
Comment (reyang) [inline releases/new-telemetry-release-process-template.md:46]: Why API but not SDK (and other packages)?

File: releases/new-telemetry-release-process-template.md (hunk 3 above)
Comment (bogdandrutu) [reply releases/new-telemetry-release-process-template.md:5]: If we find big issues (probably will not happen), then we can end-up merging it as REJECTED.

File: releases/new-telemetry-release-process.md [hunk 8]
```diff
@@ -0,0 +1,44 @@
+# New Telemetry Release Process
```
Comment (reyang) [inline releases/new-telemetry-release-process.md:1]: Should we stick with the term ["signal"](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/glossary.md#signals)?

File: releases/new-telemetry-release-process-template.md (hunk 5 above)
Comment (bogdandrutu) [reply releases/new-telemetry-release-process-template.md:11]: If it uses the same document (probably most of the time) then will be the same link.

**Review by reyang — DISMISSED** [review]
Most of my comments are non-blocking.

One blocker I'm seeing - in this PR we've introduced new terms without definition, and some of them seem to be different from what we've already defined - e.g. `telemetry type` vs. `signal`.

File: releases/new-telemetry-release-process-template.md (hunk 3 above)
Comment (reyang) [reply releases/new-telemetry-release-process-template.md:5]: I don't see why we want to merge it as REJECTED, would you help to explain?

**Comment by bogdandrutu** [comment]
@reyang have we documented `telemetry type vs. signal` somewhere, if yes happy to change to `signal` everywhere

File: releases/new-telemetry-release-process-template.md (hunk 3 above)
Comment (bogdandrutu) [reply releases/new-telemetry-release-process-template.md:5]: Sure, we can do it binary, if it is merged is approved otherwise is not merged. We do have the same thing in otep process but we can remove status.

**Comment by bogdandrutu** [comment]
Found we have signal defined, will change

File: releases/new-telemetry-release-process-template.md (hunk 2 above)
Comment (bogdandrutu) [reply releases/new-telemetry-release-process-template.md:35]: @lzchen that is our hope, but we will probably learn more during the first few reviews.

File: releases/new-telemetry-release-process-template.md (hunk 6 above)
Comment (bogdandrutu) [reply releases/new-telemetry-release-process-template.md]: Open multiple reviews?

File: releases/new-telemetry-release-process-template.md (hunk 4 above)
Comment (bogdandrutu) [reply releases/new-telemetry-release-process-template.md]: Done.

File: releases/new-telemetry-release-process-template.md (hunk 7 above)
Comment (bogdandrutu) [reply releases/new-telemetry-release-process-template.md]: Done.

File: releases/new-telemetry-release-process.md (hunk 8 above)
Comment (bogdandrutu) [reply releases/new-telemetry-release-process.md]: Sure.

File: releases/new-telemetry-release-process.md (hunk 1 above)
Comment (bogdandrutu) [reply releases/new-telemetry-release-process.md]: `technical-specification-review-python-trace.md` is the new recommended name. Probably we should include the version if a major version bump happens

Comment (bogdandrutu) [reply releases/new-telemetry-release-process.md]: Done.

**Comment by bogdandrutu** [comment]
@alolita 

> Add an overall purpose statement for this pre-release review process - what does this process aim to accomplish, outline benefits to the project.

From @tigrannajaryan :

A few reasons from the top of my head:

To help language SIGs to find out whether they are ready for GA.
To independently certify that language implementation is spec-compliant.
To ensure uniform experience across languages (spec is useful, but does not fully capture usability aspects and how the API compares for different languages).
To uncover gaps in the implementation (additional untrained pair of eyes often helps).
There is probably more.

> Add a checklist or list of questions maintainers can use for a spec compliance review.

There is a TODOs section in the template

> Add a table to outline process milestones and timelines.

Not sure I can think of how this would look like, can you provide a small example

File: releases/new-signal-release-process.md [hunk 9]
```diff
@@ -0,0 +1,44 @@
...
+
+The OpenTelemetry [Technical Committee](https://github.com/open-telemetry/community/blob/main/tech-committee-charter.md)
+commits to ensure that every decision will be made in less than 14 days after
+the process is started.
+
+To begin the technical and specification compliance review language maintainers
+MUST submit a PR and complete this [template](new-signal-release-process-template.md).
+The name of the file MUST be `release-review-${language}-${signal}-v${major_version}.md`.
+
+During the review process, breaking changes MAY be needed to comply with the
```
Comment (jsuereth) [inline releases/new-signal-release-process.md:38]: nit: `breaking changes related to the signal`.

File: releases/new-signal-release-process-template.md [hunk 10]
```diff
@@ -0,0 +1,46 @@
...
+time in a reasonable amount of time to not delay the review process.
+
+It is highly recommended to have at least 2 members of the Technical Committee as sponsors. Any Technical Committee can be a sponsor, and any Technical Committee member can provide feedback during the review process.
+
+Technical Committee Sponsor(s) MUST:
+
+* Do the due diligence and review the public documentation and examples, and
+ensure specification conformance.
+* Ensure conformance with the versioning and stability document.
+* Ensure consistent names across implementations (e.g. TraceId vs Id)
```
[the remark this answers is not in the document]
Comment (tigrannajaryan) [reply releases/new-signal-release-process-template.md:42]: I think it is will be a delicate balance. Try to be consistent but also try to avoid coming up with solutions that are not idiomatic for the language.

Very subjective, I know. :-)

File: releases/new-signal-release-process-template.md [hunk 11]
```diff
@@ -0,0 +1,46 @@
...
+* Avoid confusions across implementation (e.g same function returns has different behaviors)
```
[the remark this answers is not in the document]
Comment (tigrannajaryan) [reply releases/new-signal-release-process-template.md:43]: Maintainers usually know everything about the one language they work on, but not much about other language implementations. I'd rather burden TC than maintainers with this job.

**Comment by bogdandrutu** [comment]
For all maintiners, please do not release a stable version until a decision is made on this PR:

@open-telemetry/cpp-maintainers

@open-telemetry/go-maintainers

@open-telemetry/dotnet-maintainers

@open-telemetry/erlang-maintainers

@open-telemetry/java-maintainers

@open-telemetry/javascript-maintainers

@open-telemetry/php-maintainers  

@open-telemetry/python-maintainers 

@open-telemetry/ruby-maintainers 

@open-telemetry/swift-maintainers

File: releases/new-signal-release-process.md [hunk 12]
```diff
@@ -0,0 +1,44 @@
...
+
+In this document, the terms "OpenTelemetry" and "language implementations" both
+specifically refer to the OpenTelemetry client libraries. These terms do not
+refer to the specification or the Collector in this document.
+
+Every OpenTelmetry official major release MUST follow the release process
+proposed in this document. A  "major release" refers to a new version of any
+[signal](../specification/glossary.md#signals) (adding a new signal to an
+already released major version of the library, or a major version bump for an
+existing signal release) for OpenTelemetry client libraries.
```
Comment (aabmass) [inline releases/new-signal-release-process.md:15]:
```suggestion
existing released signal) for OpenTelemetry client libraries.
```

File: releases/new-signal-release-process-template.md [hunk 13]
```diff
@@ -0,0 +1,46 @@
+# Technical Review and Specification Conformance for ${Language}/${SIGNAL}/${MAJOR_VERSION} library
```
Comment (aabmass) [inline releases/new-signal-release-process-template.md:1]: It would be good to indicate the released client library version somewhere in the template, since it may not be the same as `${MAJOR_VERSION}`

Comment (bogdandrutu) [reply releases/new-signal-release-process-template.md:1]: A signal is only one added per major version, maybe in the section of TODOs to add that

File: releases/new-signal-release-process-template.md [hunk 14]
```diff
@@ -0,0 +1,46 @@
...
+
+Technical Committee Sponsor(s): []
+
+Expected Release Date: []
+
+Spec Compliance Matrix is up to date: [Permanent link to the matrix]
+
+Versioning And Stability Document: [LINK to version document]
```
Comment (reyang) [inline releases/new-signal-release-process-template.md:9]:
```suggestion
Versioning And Stability Document: [Link to version document]
```

Comment (bogdandrutu) [reply releases/new-signal-release-process-template.md:9]: Done.

File: releases/new-signal-release-process-template.md (hunk 13 above)
Comment (bogdandrutu) [reply releases/new-signal-release-process-template.md:1]: Done.

File: releases/new-signal-release-process.md (hunk 12 above)
Comment (bogdandrutu) [reply releases/new-signal-release-process.md:15]: Done.

File: releases/new-signal-release-process.md (hunk 9 above)
Comment (bogdandrutu) [reply releases/new-signal-release-process.md]: Done.

File: releases/new-signal-release-process-template.md [hunk 15]
```diff
@@ -0,0 +1,53 @@
+# Technical Review and Specification Conformance for ${Language}/${SIGNAL}/${MAJOR_VERSION} library
+
+Technical Committee Sponsor(s): []
+
+Expected Release Date: []
+
+Expected Version Number: []
```
Comment (reyang) [inline releases/new-signal-release-process-template.md:7]: Do we want folks to list the components that are to be released (e.g. SDK, Zipkin exporter, OpenTracing Shim) or we should review "everything" in the repo? What if the new release has significant changes in one component, but only version bump in another component - do we review both of them or just the one that has the actual change?

Comment (bogdandrutu) [reply releases/new-signal-release-process-template.md:7]: Just the one that changed

**Comment by cijothomas** [comment]
> For all maintiners, please do not release a stable version until a decision is made on this PR:
> 
> @open-telemetry/cpp-maintainers
> 
> @open-telemetry/go-maintainers
> 
> @open-telemetry/dotnet-maintainers
> 
> @open-telemetry/erlang-maintainers
> 
> @open-telemetry/java-maintainers
> 
> @open-telemetry/javascript-maintainers
> 
> @open-telemetry/php-maintainers
> 
> @open-telemetry/python-maintainers
> 
> @open-telemetry/ruby-maintainers
> 
> @open-telemetry/rust-maintainers
> 
> @open-telemetry/swift-mainteiners

@bogdandrutu This requirement conflicts with the decision made with TC members (Carlos, Sergey) and .NET Maintainers at the .NET SIG meeting on 2/9, where it was decided to release 1.0 once https://github.com/open-telemetry/opentelemetry-specification/pull/1372 is merged.

https://github.com/open-telemetry/opentelemetry-dotnet/pull/1826

File: releases/new-signal-release-process-template.md (hunk 15 above)
Comment (bogdandrutu) [reply releases/new-signal-release-process-template.md:7]: I think it is mentioned in the main document:

>Every OpenTelmetry official major release MUST follow the release process
proposed in this document. A  "major release" refers to a new version of any
[signal](../specification/glossary.md#signals) (adding a new signal to an
already released major version of the library, or a major version bump for an
existing released signal) for OpenTelemetry client libraries.

**Comment by bogdandrutu** [comment]
@cijothomas I think during the maintainers meeting we talked about this and decided for everyone to wait until Wed for this to be published and reviewed. I don't think we asked for postponing too much time but here we are, too late to complain anyway.

**Comment by bogdandrutu** [comment]
@reyang I think you can dismiss the request changes because the main concern was addressed, unless you have other big concerns

File: releases/new-telemetry-release-process-template.md (hunk 2 above)
Comment (lzchen) [reply releases/new-telemetry-release-process-template.md]: @bogdandrutu 
I asked for clarification between two options. Could you clarify which one it is? xD

File: releases/new-signal-release-process.md [hunk 16]
```diff
@@ -0,0 +1,44 @@
...
+
+This document defines the release requirements and process followed by the
+OpenTelemetry libraries, along with the rules and procedures for meeting those
+requirements.
+
+In this document, the terms "OpenTelemetry" and "language implementations" both
+specifically refer to the OpenTelemetry client libraries. These terms do not
+refer to the specification or the Collector in this document.
+
+Every OpenTelmetry official major release MUST follow the release process
```
Comment (cijothomas) [inline releases/new-signal-release-process.md:11]:
```suggestion
Every OpenTelemetry official major release MUST follow the release process
```

Comment (bogdandrutu) [reply releases/new-signal-release-process.md]: done.

File: releases/new-signal-release-process-template.md (hunk 11 above)
[the remark this answers is not in the document]
Comment (jsuereth) [reply releases/new-signal-release-process-template.md:48]: AH, I misread this as implementation differences within an SDK.

I assume, though, some leniency is allowed here (e.g. Java's Completable-Returns vs. other languages)?

[the remark this answers is not in the document]
Comment (bogdandrutu) [reply releases/new-signal-release-process-template.md:48]: Yes :) we are not robots :))

[the remark this answers is not in the document]
Comment (yurishkuro) [reply releases/new-signal-release-process-template.md:48]: > we are not robots

speak for yourself

File: releases/new-telemetry-release-process-template.md (hunk 2 above)
Comment (bogdandrutu) [reply releases/new-telemetry-release-process-template.md]: TC aims and will put effort into finishing the entire process in 14 days, for this language maintainers need to be responsive :) I think this is how I would rephrase everything

**Review by ramthi — CHANGES_REQUESTED** [review]

File: releases/new-signal-release-process-template.md [hunk 17]
```diff
@@ -0,0 +1,53 @@
...
+[the versioning and stability requirements](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/versioning-and-stability.md)
+and prepare the public documentation about the versioning and stability guarantees.
+* Work with a Technical Committee Sponsor(s) during review. The Technical
+Committee members are not language experts, and it is expected to work with the
+language maintainers and/or invited language experts to perform this review
+process. Language maintainers SHOULD respond to any question during the review
+time in a reasonable amount of time to not delay the review process.
+
+It is highly recommended to have at least 2 members of the Technical Committee
+as sponsors. Any Technical Committee can be a sponsor, and any Technical
```
Comment (lalitb) [inline releases/new-signal-release-process-template.md:39]:
```suggestion
as sponsors. Any Technical Committee member can be a sponsor, and any Technical
```

File: releases/new-signal-release-process-template.md [hunk 18]
```diff
@@ -0,0 +1,53 @@
...
+Committee member can provide feedback during the review process.
+
+Technical Committee Sponsor(s) MUST:
+
+* Do the due diligence and review the public documentation and examples, and
+ensure specification conformance.
+* Ensure conformance with the versioning and stability document.
+* Ensure consistent names across implementations (e.g. TraceId vs GlobalId)
+* Avoid confusions across implementation (e.g same function returns has different behaviors)
+* Ensure no experimental features (signals) are part of the released packages (e.g. api, sdk, etc.).
```
Comment (lalitb) [inline releases/new-signal-release-process-template.md:49]: For opentelemetry-cpp, the releases are the source distributions, and the experimental (unstable) telemetry signals and features will be introduced behind feature flag protected by a preprocessor macro. So though metrics and logging signal would be distributed, they wouldn't be included in builds unless the feature flag is explicitly enabled. This has been clarified in our versioning doc too ( https://github.com/open-telemetry/opentelemetry-cpp/blob/main/Versioning.md )

File: releases/new-signal-release-process-template.md (hunk 11 above)
[the remark this answers is not in the document]
Comment (lalitb) [reply releases/new-signal-release-process-template.md:48]: Not sure if I understand this clearly. Functions are internal to language implementation, and it is possible for them to behave differently across language implementations, and should be fine as long as external api/sdk interfaces are consistent across languages, and compliant with specs.

File: releases/new-signal-release-process.md [hunk 19]
```diff
@@ -0,0 +1,44 @@
...
+
+* Technical and Specification Compliance Review
+
+The OpenTelemetry Technical Committee reserves the right to change and improve
+this process during time based on the prior experiences.
+
+## Technical and Specification Compliance Review
+
+The OpenTelemetry [Technical Committee](https://github.com/open-telemetry/community/blob/main/tech-committee-charter.md)
+commits to ensure that every decision will be made in less than 14 days after
```
Comment (lalitb) [inline releases/new-signal-release-process.md:31]: Will it be 14 days ( 2 weeks ) or 14 weekdays/working days (~3 weeks )?

File: releases/new-signal-release-process-template.md (hunk 18 above)
Comment (reyang) [reply releases/new-signal-release-process-template.md:49]: @open-telemetry/cpp-maintainers

File: releases/new-signal-release-process-template.md [hunk 20]
```diff
@@ -0,0 +1,53 @@
...
+The Language Maintainers MUST:
+
+* Prior to starting the review, update the
+[compliance matrix](https://github.com/open-telemetry/opentelemetry-specification/blob/main/spec-compliance-matrix.md)
+for that language/signal.
+* Each language implementation MUST follow
+[the versioning and stability requirements](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/versioning-and-stability.md)
+and prepare the public documentation about the versioning and stability guarantees.
+* Work with a Technical Committee Sponsor(s) during review. The Technical
+Committee members are not language experts, and it is expected to work with the
```
Comment (reyang) [inline releases/new-signal-release-process-template.md:33]:
```suggestion
Committee members are not language experts, and are expected to work with the
```

File: releases/new-signal-release-process-template.md [hunk 21]
```diff
@@ -0,0 +1,53 @@
...
+language maintainers and/or invited language experts to perform this review
+process. Language maintainers SHOULD respond to any question during the review
```
Comment (reyang) [inline releases/new-signal-release-process-template.md:35]: It is hard to understand what "any question" means here, and how the communication will be done (e.g. private chat, public meeting with recording enabled, GitHub issues, Google doc).

I suggest that we make it explicit - all the communications from the TC members (and invited language experts if there is any) to the maintainers will be summarized and published as GitHub issues against the language implementation repo, with a top level issue which contains the links to individual issues.

File: releases/new-signal-release-process-template.md [hunk 22]
```diff
@@ -0,0 +1,53 @@
...
+time in a reasonable amount of time to not delay the review process.
```
Comment (reyang) [inline releases/new-signal-release-process-template.md:36]: It is unclear what is the common understanding of "reasonable amount of time", probably just put something like "two working days" similar like what we did [here](https://github.com/open-telemetry/opentelemetry-specification/blob/main/CONTRIBUTING.md#how-to-get-your-pr-merged)?

File: releases/new-signal-release-process.md [hunk 23]
```diff
@@ -0,0 +1,44 @@
...
+the process is started.
+
+To begin the technical and specification compliance review language maintainers
+MUST submit a PR and complete this [template](new-signal-release-process-template.md).
```
Comment (reyang) [inline releases/new-signal-release-process.md:35]: This will have potential conflict with the [current process](https://github.com/open-telemetry/opentelemetry-specification/blob/main/CONTRIBUTING.md#how-to-get-your-pr-merged), which mentioned that we require at minimum two valid approvals + two working days. If a "release-review-xyz-metrics-v1.0.md" PR was sent and got two valid approvals and merged by the owner, we still should not merge it if we follow this doc.

Comment (reyang) [inline releases/new-signal-release-process.md:35]: Minor, consider putting a recommendation on the PR title (or having a PR template and put a link here).

File: releases/new-signal-release-process.md [hunk 24]
```diff
@@ -0,0 +1,44 @@
...
+The name of the file MUST be `release-review-${language}-${signal}-v${major_version}.md`.
+
+During the review process, breaking changes related to the new signal MAY be
+needed to comply with the specification, it is highly recommended to start
+the process as soon as possible, but after all features and planned changes are merged.
```
Comment (reyang) [inline releases/new-signal-release-process.md:40]: If I understand it correctly, we encourage people to do this right before they move from beta to rc (release candidate)?
Or the actual recommendation is to go through the process when they want to go from rc to stable release?

File: releases/new-signal-release-process.md [hunk 25]
```diff
@@ -0,0 +1,44 @@
...
+
+After the review is approved, it is highly recommended to not make any major
```
Comment (reyang) [inline releases/new-signal-release-process.md:42]: it is very vague what "major change" (e.g. is moving one API from one namespace to another considered as major change? is package name change considered as major change?) - can we require each client to implement public API enforcement rule as part of CI. In this way we can make it clear that "if any public API is changed, it must go through the process again".

File: releases/new-signal-release-process-template.md [hunk 26]
```diff
@@ -0,0 +1,53 @@
+# Technical Review and Specification Conformance for ${Language}/${SIGNAL}/${MAJOR_VERSION} library
```
Comment (reyang) [inline releases/new-signal-release-process-template.md:1]:
```suggestion
# Technical Review and Specification Conformance for ${LANGUAGE}/${SIGNAL}/${MAJOR_VERSION} library
```

File: releases/new-telemetry-release-process-template.md (hunk 2 above)
Comment (reyang) [reply releases/new-telemetry-release-process-template.md]: > @lzchen that is our hope, but we will probably learn more during the first few reviews.

If the goal is to learn by trying out the process with few reviews, we should go through OTEP first.

File: releases/new-signal-release-process-template.md (hunk 18 above)
Comment (tigrannajaryan) [reply releases/new-signal-release-process-template.md:49]: Hiding experimental features behind feature flags is a great approach. It lets you make progress, yet prevents exposing half-baked features to end users. We should encourage all SIGs to take this approach.

File: releases/new-telemetry-release-process-template.md (hunk 2 above)
Comment (bogdandrutu) [reply releases/new-telemetry-release-process-template.md]: OTEP of what?

File: releases/new-signal-release-process-template.md (hunk 11 above)
[the remark this answers is not in the document]
Comment (bogdandrutu) [reply releases/new-signal-release-process-template.md:48]: @lalitb we are only talking about public API here. I will clarify.

File: releases/new-signal-release-process-template.md (hunk 17 above)
Comment (bogdandrutu) [reply releases/new-signal-release-process-template.md:39]: Done.

File: releases/new-signal-release-process-template.md (hunk 11 above)
[the remark this answers is not in the document]
Comment (bogdandrutu) [reply releases/new-signal-release-process-template.md:48]: I think now is clear, please open if not.

File: releases/new-signal-release-process.md (hunk 19 above)
Comment (bogdandrutu) [reply releases/new-signal-release-process.md:31]: 2 weeks is the wish

File: releases/new-signal-release-process.md (hunk 23 above)
Comment (bogdandrutu) [reply releases/new-signal-release-process.md:35]: Added a statement that this is an exception from the rule.

File: releases/new-signal-release-process-template.md (hunk 26 above)
Comment (bogdandrutu) [reply releases/new-signal-release-process-template.md:1]: done.

File: releases/new-signal-release-process.md (hunk 23 above)
Comment (bogdandrutu) [reply releases/new-signal-release-process.md:35]: done.

File: releases/new-signal-release-process-template.md (hunk 18 above)
Comment (bogdandrutu) [reply releases/new-signal-release-process-template.md:49]: I think there are multiple ways to ensure experimental features are not public. So this can be one of them.

File: releases/new-signal-release-process-template.md [hunk 27]
```diff
@@ -0,0 +1,53 @@
...
+
+Technical Committee Sponsor(s): []
+
+Expected Release Date: []
+
+Expected Version Number: []
+
+Spec Compliance Matrix is up to date: [Permanent link to the matrix]
```
[the remark this answers is not in the document]
Comment (bogdandrutu) [reply releases/new-signal-release-process-template.md:9]: That is a possibility, let's not over specified. I think any link that proves and shows the current status of the matrix is good :)

File: releases/new-signal-release-process-template.md (hunk 20 above)
Comment (bogdandrutu) [reply releases/new-signal-release-process-template.md:33]: done.

File: releases/new-signal-release-process-template.md (hunk 22 above)
Comment (bogdandrutu) [reply releases/new-signal-release-process-template.md:36]: sure, added "ideally 1 business day".

File: releases/new-signal-release-process-template.md (hunk 21 above)
Comment (bogdandrutu) [reply releases/new-signal-release-process-template.md:35]: Done. clarify the communication channels

File: releases/new-signal-release-process.md (hunk 24 above)
Comment (bogdandrutu) [reply releases/new-signal-release-process.md:41]: done.

File: releases/new-signal-release-process.md (hunk 25 above)
Comment (bogdandrutu) [reply releases/new-signal-release-process.md:44]: if you rename the package from `opentelemetry.trace` to `opentelemetry.foobar` I would say yes :).

File: releases/new-telemetry-release-process-template.md (hunk 2 above)
Comment (reyang) [reply releases/new-telemetry-release-process-template.md]: I suggest that we pick two language clients that are willing to try out this process, go through the steps proposed in this PR until they release a stable version. After two language clients released, we can take the learning/feedback and see whether we should move forward or back out.

Whether this PR should be here or somewhere else (e.g. in the community repo) - I don't know, probably not here as it is weird to put some process document and request form in a spec repo (unless the process is about the spec itself). No strong opinion though.

**Comment by iNikem** [comment]
Should autoinstrumentation repos follow the same review process?

File: releases/new-telemetry-release-process-template.md (hunk 2 above)
Comment (bogdandrutu) [reply releases/new-telemetry-release-process-template.md]: that is what we will do naturally, not all the languages are ready for this. So not sure if anything is needed to be said here.

**Comment by bogdandrutu** [comment]
@open-telemetry/technical-committee please review, this blocks progress if we do not merge or reject this proposal.

**Review by yurishkuro — COMMENTED** [review]
I am in favor of the formalized process with a checklist to follow / template doc to fill out. But I'm not as certain that the TC must always be involved. We're operating on the honor system. A language team could just release a new version without following the process. If they do follow the process then I would rather trust them to follow it all the way, without an extra friction of involving TC.

**Comment by bogdandrutu** [comment]
@yurishkuro point taken. But I think it is hard to come with that list with going through the process couple of times to identify all the requirements.

Alternatively I can see us just requiring the checklist and as a best effort TC members do what I suggest here once or twice and bring the learnings to the process back

**Comment by yurishkuro** [comment]
> it is hard to come with that list 

That's fair, we can certainly do a few test runs, especially since many TC members are actually involved in some of the SIGs. I would just prefer to take it easy on the MUST wording for TC involvement (I am ok with MUST for the process itself).

**Review by carlosalberto — APPROVED** [review]

**Comment by carlosalberto** [comment]
LGTM - as there are still standing questions (and doubts), I suggest we give this a shot to a few **volunteering** SIGs and document our findings/feedback in further PRs to tune this process.

File: CONTRIBUTING.md [hunk 28]
```diff
@@ -193,3 +193,6 @@ on each other), the owner should try to get people aligned by:
 If none of the above worked and the PR has been stuck for more than 2 weeks, the
 owner should bring it to the [OpenTelemetry Specification SIG
 meeting](https://github.com/open-telemetry/community#cross-language-specification).
+
+New signal release reviews are subject to different rules that and do not follow
```
Comment (ThomsonTan) [inline CONTRIBUTING.md:197]:
```suggestion
New signal release reviews are subject to different rules and do not follow
```

File: releases/new-signal-release-process-template.md [hunk 29]
```diff
@@ -0,0 +1,55 @@
...
+as sponsors. Any Technical Committee member can be a sponsor, and any Technical
+Committee member can provide feedback during the review process.
+
+Technical Committee Sponsor(s) MUST:
+
+* Do the due diligence and review the public documentation and examples, and
+ensure specification conformance.
+* Ensure conformance with the versioning and stability document.
+* Ensure consistent names across implementations (e.g. TraceId vs GlobalId)
+* Avoid confusions across implementation (e.g same public API has different behaviors)
```
Comment (ThomsonTan) [inline releases/new-signal-release-process-template.md:50]:
```suggestion
* Avoid confusions across implementation (e.g. same public API has different behaviors)
```

File: releases/new-signal-release-process-template.md (hunk 18 above)
Comment (ThomsonTan) [reply releases/new-signal-release-process-template.md:51]: As we could include experimental feature in the released package, should we say "Ensure no experimental features/signals are enabled by default in the released packages?

**Review by tigrannajaryan — APPROVED** [review]
LGTM.

To make it clear: for me this process is a way for TC members to help language SIGs to make a release, to help ensure the language implementation is consistent with other languages, confirms with the spec. We want the TC member to be another pair of fresh eyes who can help the SIG to go over the release checklist.
 
There is absolutely no intent to introduce a process that "controls" the language SIGs or introduces obstacles in the release process. If any of the language SIGs notice that the process is not helpful the TC would like to know about it immediately so that we can fix (or eliminate) the process.

**Comment by lzchen** [comment]
@tigrannajaryan 
The Python SIG has been trying this out in a non-formal way. @carlosalberto has volunteered to go through our API/SDK surface and review it thoroughly for discrepancies, and as a maintainer I find this to be of great value.

Personally, I find the wording of the PR to be quite aggressive and way more formal than it needs to be. Having to go through a process just adds more overheads onto the maintainers and TC member(s) responsible for reviewing. I am totally fine with working offline directly with a TC member as a maintainer to discuss any of the logistics, but a whole process seems to be overkill.

The only benefit I see from having a process is that it will encourage reviewers/approvers/maintainers/TC members to prioritize the tasks outlined in the "review" process as something to be done as soon as possible. The two week timeline "lights a fire" under contributors and I think that is valuable (it certainly caused Python contributors to prioritize a lot of the tasks that @carlosalberto had found). Just my thoughts from the first few days of going through this process "informally".

File: releases/new-signal-release-process-template.md (hunk 27 above)
[the remark this answers is not in the document]
Comment (dyladan) [reply releases/new-signal-release-process-template.md:9]: Link to the latest matrix or the matrix commit at the time of the review?

**Comment by carlosalberto** [comment]
After trying this out with the Python SIG, I suggest with this along the following points:

* Define this as general review process that SIGs can opt in.
* The issues and changes that the reviewer points out can be further discussed with the maintainers, and they MAY be applied or not, at the discretion of the maintainers themselves.

@bogdandrutu Sounds like a good compromise? And in general, having this document as a general starting point is something really valuable to have, even if not _all_ SIGs end up doing it.
